qutip/core/cy/cqobjevo.pyx

```diff
@@ -32,1189 +34,183 @@
-            sparse_list.append(op[0].data)
-        sparse_list += [cte.data]
-        matched = _zcsr_match(sparse_list)
-
-        self.indptr = matched[0].indptr
-        self.indices = matched[0].indices
-        self.cte = matched[-1].data
-        self.nnz = len(self.cte)
-        self.data_t = np.zeros((self.nnz), dtype=complex)
-        self.data_ptr = &self.data_t[0]
-
-        self.ops = np.zeros((self.num_ops, self.nnz), dtype=complex)
-        for i, op in enumerate(matched[:-1]):
-          for j in range(self.nnz):
-            self.ops[i,j] = op.data[j]
-
-    def set_factor(self, func=None, ptr=False, obj=None):
-        self.factor_use_cobj = 0
-        if func is not None:
-            self.factor_func = func
-        elif obj is not None:
-            self.factor_use_cobj = 1
-            self.factor_cobj = obj
-        else:
-            raise Exception("Could not set coefficient function")
-
-    def __getstate__(self):
-        return (self.shape0, self.shape1, self.dims, self.nnz, self.super,
-                self.factor_use_cobj, self.factor_cobj, self.factor_func,
-                self.dyn_args, self.num_ops,
-                np.array(self.indptr), np.array(self.indices),
-                np.array(self.cte), np.array(self.ops))
-
-    def __setstate__(self, state):
-        self.shape0 = state[0]
-        self.shape1 = state[1]
-        self.dims = state[2]
-        self.nnz = state[3]
-        self.super = state[4]
-        self.factor_use_cobj = state[5]
-        if self.factor_use_cobj:
-            self.factor_cobj = <CoeffFunc> state[6]
-        self.factor_func = state[7]
-        self.dyn_args = state[8]
-        self.num_ops = state[9]
-        self.indptr = state[10]
-        self.indices = state[11]
-        self.cte = state[12]
-        self.ops = state[13]
-        self.coeff = np.zeros((self.num_ops), dtype=complex)
-        self.coeff_ptr = &self.coeff[0]
-        self.data_t = np.zeros((self.nnz), dtype=complex)
-        self.data_ptr = &self.data_t[0]
-
-    @cython.boundscheck(False)
-    @cython.wraparound(False)
-    @cython.cdivision(True)
-    cdef void _call_core(self, complex[::1] out, complex* coeff):
-        cdef int i, j
-        cdef complex * ptr
-        ptr = &self.cte[0]
-        for j in range(self.nnz):
-            out[j] = ptr[j]
-        for i in range(self.num_ops):
-            ptr = &self.ops[i,0]
-            for j in range(self.nnz):
-                out[j] += ptr[j] * coeff[i]
+            coeff = self.factor_func(t)
+            PyErr_CheckSignals()
+            for i in range(self.n_ops):
+                self.coefficients[i] = coeff[i]
+        return
 
-    def call(self, double t, int data=0):
-        cdef int i
-        cdef complex[::1] data_t = np.empty(self.nnz, dtype=complex)
+    cdef void _factor_dynamic(self, double t, Data state) except *:
+        # TODO: why are `dynamic` arguments here when they're purely the
+        # concern of the solvers?
```

Comment:
QobjEvo are made to remove code duplication in solver, they are not used much anywhere else and don't have much user facing documentation. So it is/were part of the solver not the core.
By making it an independent, user-facing object in core, the solver specific features will need to be moved in a new object.

For now, solvers will be shaky since QobjEvo no longer fit their specific needs.